### ISSUE TITLE
Switch to Azure pipelines for CI

### ISSUE THREAD

**Issue by StarpTech** [fastify/fastify#1218]
## 🚀 Feature Proposal

We should switch to Azure pipelines for CI

## Motivation

* Free for opensource with 10 Parallel Jobs
* Support for Mac, Windows and Linux
* Easy installation as with Travis
* Much better configuration and debugging experience
* Full Github integration
* Cross platform builds are much faster than in travis right now.

## Example

![image](https://user-images.githubusercontent.com/1764424/47241255-53b5d180-d3eb-11e8-9d7e-f9d6d800f144.png)

Example configuration: https://github.com/Prettyhtml/prettyhtml/blob/master/azure-pipelines.yml

**Comment by cemremengu** [comment]
How is the integration with other services like `coveralls` for example? I am personally a bit negatively biased against microsoft products for these kind of things but nevertheless I am neutral against the change

**Comment by StarpTech** [comment]
@cemremengu there exists no "native" coveralls integration on ci level. You only need to run another command. You know that Microsoft owns Github right? 😄

**Comment by StarpTech** [comment]
> AND does it have a badge ? 😄

Yes

**Comment by cemremengu** [comment]
Yes and I fear that one day we will wake up to `Github 365` 😄 

Joking aside, considering that and the future, it may be a good idea to switch though. I think MSFT products will integrate much better with Github as you suggested.

**Comment by StarpTech** [comment]
It's for free! and it will make our CI much simpler and more transparently. I don't care whom it belongs. Fastify is 100% opensource nothing to care about.

**Comment by mcollina** [comment]
As I said elsewhere, I’m open for experimentation if it means we can provide a better experience for our contributors. I’ll see if I can enable it next week.

**Comment by kaylangan** [comment]
I'm a Program Manager on Azure Pipelines. Let me know if you have any questions or suggestions. I'm happy to help and excited that you're considering trying us out!

**Comment by mcollina** [comment]
I'll be happy to test this out and see how this goes alongside what we already have. A single CI for Windows/Linux/Mac is what we need really.

WDYT @delvedor? 

@kaylangan is there any specific steps we would have to do to add this? Should I just add it from the Github marketplace?

**Comment by cemremengu** [comment]
@kaylangan I took a quick look but couldn't see any simple way to setup services like in travis:

```
services
 - mongodb
```

looks like we need a dockerfile for this kind of things? Another problem was that I have to logout from my company's microsoft account to login to devops :)

**Comment by delvedor** [comment]
I agree that a single CI for Windows/Linux/Mac would be awesome.
But we should also consider how is easy to use for external contributors, the thing that I most love about Travis, is that the user experience is excellent!
I will give this a try soon :)

**Comment by jsumners** [comment]
I agree with @cemremengu. I'm not very pleased that I have to have a unique Microsoft account to use this service. Travis just uses GitHub's OAuth provider which makes things easy.

**Comment by mcollina** [comment]
Some notes:

- It seems to be extremely better than AppVeyor for WIndows. I'm not sure about Travis, but it seems way faster.
- Can you access/configure the CI? Who has a microsoft account that I could try to add, or are permissions managed automatically?
- I cannot configure Travis from my account anymore due to a bug (ETOOMANY orgs) in GitHub itself and the new marketplace integration. I hope they solve this soon. Azure Pipelines somehow works as it uses some different forms.

@kaylangan I think it would be awesome to login via github integration on Azure pipelines. Is it something that is in your roadmap?

**Comment by kaylangan** [comment]
Sorry for the delay. It looks like you were able to get it set up--the GitHub app/marketplace are both good options.

We are currently working on integrating with GitHub identities, but you are correct--that support does not yet exist. For now, you need an MS identity to first set up the pipeline. But after it's been set up, anyone can view build results publicly without an MS identity. Additionally, because your pipeline is configured as code, anyone can make updates to the build definition via PR.

I'll need to get back to you on setting up services--it's not quite my wheelhouse.

**Comment by kaylangan** [comment]
> I took a quick look but couldn't see any simple way to setup services as easy as in travis:
> 
> ```
> services
>  - mongodb
> ```
> looks like we need a dockerfile for this kind of things?

Just to close the loop on this. We're currently working on the syntax for this. In the meantime, yes, you can `docker run` your own stuff.
